crates/metadata-store/src/rocksdb/server.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use bytes::Bytes;
+use bytestring::ByteString;
+use restate_core::cancellation_watcher;
+use restate_core::network::{MessageHandler, NetworkSender};
+use restate_node_protocol::metadata_store::{
+    Delete, Get, GetVersion, MetadataStoreRequest, MetadataStoreResponse, Put, RequestKind,
+    ResponseKind,
+};
+use restate_node_protocol::{metadata_store, MessageEnvelope};
+use restate_types::errors::GenericError;
+use restate_types::metadata_store::{Precondition, VersionedValue};
+use restate_types::{GenerationalNodeId, NodeId, Version};
+use rocksdb::{ColumnFamily, Options, WriteOptions, DB};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::path::Path;
+use tokio::sync::mpsc;
+use tracing::{debug, trace};
+
+// todo: Introduce bound if this channel should grow unboundedly
+type CommandSender = mpsc::UnboundedSender<(GenerationalNodeId, MetadataStoreRequest)>;
+type CommandReceiver = mpsc::UnboundedReceiver<(GenerationalNodeId, MetadataStoreRequest)>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("storage error: {0}")]
+    Storage(#[from] rocksdb::Error),
+    #[error("violated precondition: {0}")]
+    PreconditionViolation(String),
+    #[error("codec error: {0}")]
+    Codec(GenericError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BuildError {
+    #[error("failed opening rocksdb: {0}")]
+    RocksDB(#[from] rocksdb::Error),
+}
+
+impl Error {
+    fn kv_pair_exists() -> Self {
+        Error::PreconditionViolation("key-value pair already exists".to_owned())
+    }
+
+    fn version_mismatch(expected: &Version, actual: Option<&Version>) -> Self {
+        Error::PreconditionViolation(format!(
+            "Expected version '{}' but found version '{:?}'",
+            expected, actual
+        ))
+    }
+}
+
+impl From<Error> for metadata_store::Error {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::Storage(err) => metadata_store::Error::Internal(err.to_string()),
+            Error::PreconditionViolation(msg) => metadata_store::Error::PreconditionViolation(msg),
+            Error::Codec(err) => metadata_store::Error::Internal(err.to_string()),
+        }
+    }
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+const KV_PAIRS: &str = "KV_PAIRS";
+
+/// Single node metadata store which stores the key value pairs in RocksDB.
+///
+/// In order to avoid issues arising from concurrency, we run the metadata
+/// store in a single thread.
+pub struct RocksDBMetadataStore<N> {
```

Comment:
Sounds like a good idea. I will do `s/rocksdb/local/`.

Reply:
Technically, the metadata store is not only local. It can run remotely on a different node. So maybe the more distinguishing property would be single node/not replicated? I guess the same could apply to the loglet if it can be reached via the network.

Reply:
Good call. let's refine the naming as we progress forward. Eventually single/multinode will be merged I suppose.